src/commands/common/send-transaction-command.js

```diff
@@ -0,0 +1,130 @@
+import Command from '../command.js';
+import { KNOWN_TRANSACTION_ERRORS, OPERATION_ID_STATUS } from '../../constants/constants.js';
+
+class SendTransactionCommand extends Command {
+    async sendTransactionAndHandleResult(transactionCompletePromise, data, command) {
+        const {
+            blockchain,
+            agreementId,
+            epoch,
+            operationId,
+            closestNode,
+            leftNeighborhoodEdge,
+            rightNeighborhoodEdge,
+            contract,
+            tokenId,
+            keyword,
+            hashFunctionId,
+            stateIndex,
+            txGasPrice,
+        } = data;
+        const sendTransactionOperationId = this.operationIdService.generateId();
+        let txSuccess;
+        let msgBase;
+        try {
+            this.operationIdService.emitChangeEvent(
+                this.TX_START,
+                sendTransactionOperationId,
+                blockchain,
+                agreementId,
+                epoch,
+                operationId,
+            );
+            txSuccess = await transactionCompletePromise;
+        } catch (error) {
+            this.logger.warn(
+                `Failed to execute ${command.name}, Error Message: ${error.message} for the Service Agreement ` +
+                    `with the ID: ${agreementId}, Blockchain: ${blockchain}, Contract: ${contract}, ` +
+                    `Token ID: ${tokenId}, Keyword: ${keyword}, Hash function ID: ${hashFunctionId}, ` +
+                    `Epoch: ${epoch}, State Index: ${stateIndex}, Operation ID: ${operationId}, ` +
+                    `Closest Node: ${closestNode}, Left neighborhood edge: ${leftNeighborhoodEdge}, ` +
+                    `Right neighborhood edge: ${rightNeighborhoodEdge}, ` +
+                    `Retry number: ${this.RETRY_NUMBER - command.retries + 1}.`,
+            );
+            this.operationIdService.emitChangeEvent(
+                OPERATION_ID_STATUS.FAILED,
+                sendTransactionOperationId,
+                blockchain,
+                error.message,
+                this.TX_ERROR,
+            );
+            txSuccess = false;
+            if (error.message.includes(KNOWN_TRANSACTION_ERRORS.NODE_ALREADY_SUBMITTED_COMMIT)) {
+                msgBase = 'Node has already submitted commit. Finishing';
+            } else if (error.message.includes(KNOWN_TRANSACTION_ERRORS.NODE_ALREADY_REWARDED)) {
+                msgBase = 'Node has already sent proof. Finishing';
+            } else if (error.message.includes(KNOWN_TRANSACTION_ERRORS.INSUFFICIENT_FUNDS)) {
+                msgBase = 'Insufficient funds. Finishing';
+                if (this.insufficientFundsErrorReceived) {
+                    await this.insufficientFundsErrorReceived(command.data);
```

Comment:
Do you think it would be a good idea to have some mechanism to remove operational wallets and transaction queues when out of funds, and then add them back if node runner funds the wallets ?

Reply:
We were thinking about that approach but the issue with that is that we would need to create new command or some other mechanism to check if node runner funded the wallet. Also we didn't want to add check for funding every time node tries to send transaction since it would introduce new RPC calls and we don't want that.